### DISCUSSION TITLE
Deploying always loops back to this issue: E_INVALID_FLAG: Unknown flag "production"

### DISCUSSION THREAD

**Discussion by mckayze** [adonisjs/core#3258] [Help] [answered]
## Package version
"5.3.0"

## Node.js and npm version
14.17.0 - 8.0.0

## Sample Code (to reproduce the issue)
This issue isn't code specific it seems as though its just with deployment.

## Problem
When deploying i get the following logs which i initially thought was connected to digitalocean but when i used another service and got the same issue i figured out it may be an adonis problem.

```
[2021-10-17 22:54:49] > node ace build --production
[2021-10-17 22:54:49] 
[2021-10-17 22:54:50] [ error ]  E_INVALID_FLAG: Unknown flag "production"
[2021-10-17 22:54:50] ERROR: failed to build: exit status 1
[2021-10-17 22:54:50] building: exit status 1
[2021-10-17 22:54:56] 
[2021-10-17 22:54:56] For documentation on the buildpacks used to build your app, please see:
[2021-10-17 22:54:56]    Node.js: https://do.co/apps-buildpack-node
[2021-10-17 22:54:56] 
[2021-10-17 22:54:56]  ! Build failed (145)
```

I've switched from initially using Yarn which this issue happened with to using NPM and i still get the same error message. It sometimes happens on re-builds and not on the initial build. There doesn't seem to be any pattern (that i can recognise at least) to when it happens.

The issue seems to be that the build function isn't found `node ace build` when i replace `npm run build` with `node ace build` it says that build is not found. And logging out a `node ace` during the build process doesn't show the build command at all while locally it does.

I tried to get a solution on discord and someone mentioned that running `npm i` twice worked for them, i tried this but it didn't work with both yarn and npm.

**Comment by thetutlage** [comment]
Can you please share how do you upload your project to the production server and how do you install dependencies there?

**Comment by mckayze** [comment]
Hey thanks for replying,

I upload my project to gitlab first, then i choose the branch from DO's apps services. 
To install  / build dependencies we run `npm run build`
Afterwards we run `npm run start:production` which runs this command `node build/server.js` (this works locally for reference)

**Comment by thetutlage** [comment, marked as answer]
I still don't have clarity to the complete process. But anyways, this is happening because the `@adonisjs/assembler` is not installed on your deployment server.

The package is a development dependency and it seems like your server is installing only production dependencies. This happens when `NODE_ENV` is set to production or either you are using `--production` flag with the npm install command.

**Reply by gapgag55** [reply]
Curiousing, Why do we need to install dev dependencies on production?

**Comment by mckayze** [comment]
Okay thanks this actually clears some stuff up,I thought the whole point of dev dependencies is that you don't install them on a production server, I guess i must be missing something here then! 

I've been following this guide on the adonisjs website for deploying to DO: https://docs.adonisjs.com/cookbooks/deploy-to-digial-ocean#configuring-the-do-app

**Reply by thetutlage** [reply]
Yes. But you are building your code which needs development dependencies too.

If you push compiled code to the server, then only prod dependencies are required

**Reply by axeltaylor** [reply]
@mckayze I run into the same issue, just upload the yaml file (download from the settings) with the env variable set to:
     - key: NODE_ENV
        scope: **RUN_TIME**
        value: production

And everything should be working.

**Reply by mckayze** [reply]
Thanks for this! I think I was thinking about how to deploy this wrong in previous jobs we've uploaded the codebase and compiled it on the deploy server but this makes sense. 👌

**Comment by curious-monggo** [comment]
Hi, you can try comparing local and server adonis versions:
`node ace --version`
you might not have installed assembler

Run:
`npm i @adonisjs/assembler --save-dev`

(Context: I use an .sh file to run git pull, build, pm2. Not recommended, but works for now)

**Comment by capoia** [comment]
Already had a fix on this?
I was using what @axeltaylor suggested, changing the digitalocean spcs to RUN_TIME, solves the build not found error, but now it's generating the error that NODE_ENV is not available (after build)
DigitalOcean must have changed something.

**Comment by Asylian21** [comment]
Same issue today :/ spent whole day to fix it 😤

**Reply by DiegoFelipe** [reply]
how you did that? I have installed @adonisjs/assembler package, but I still get this error, I'm using adonis v4, ace version 4.1.0
